### PR TITLE
Introduce `DB#ExecTx()` method

### PR DESCRIPTION
**Pull request by yhabteab**
Introduces a `DB#ExecTx()` method on the `DB` receiver (cherry-picked from #18)

### CODE REVIEW

File: database/db.go [hunk 1]
```diff
@@ -763,6 +763,26 @@ func (db *DB) Delete(
...
+		return errors.Wrap(err, "DB.RunInTx: cannot start a database transaction")
+	}
+	// We don't expect meaningful errors from rolling back the tx other than the sql.ErrTxDone, so just ignore it.
+	defer func() { _ = tx.Rollback() }()
+
+	if err := fn(ctx, tx); err != nil {
+		return err
+	}
+
+	return errors.Wrap(tx.Commit(), "DB.RunInTx: cannot commit a database transaction")
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply database/db.go:783]: > Above in line 773, you have also performed an `err != nil` check first

I did add a nil check there, because if there wasn't an error, I would want to go on with other things, but in this case `tx.Commit()` is the last statement of that method, so I didn't care too much about it, since `errors.Wrap()` also supports nil errors.

**Review by oxzi — DISMISSED** [review]

File: database/db.go [hunk 2]
```diff
@@ -763,6 +763,29 @@ func (db *DB) Delete(
 	return db.DeleteStreamed(ctx, entityType, idsCh, onSuccess...)
 }
 
+// RunInTx allows running a function in a database transaction without requiring manual transaction handling.
+//
+// A new transaction is started on [DB] which is then passed to fn. After fn returns, the transaction is
+// committed unless an error was returned. If fn returns an error, that error is returned or when failing
+// to start or/and commit the transaction.
+func (db *DB) RunInTx(ctx context.Context, fn func(context.Context, *sqlx.Tx) error) error {
```
Comment (lippserd) [inline database/db.go:771]: I would rename that function to `ExecTx`. Isolation level should also be adjustable. Please also consider that we already have special handling for serializable transactions.

File: database/db.go [hunk 3]
```diff
@@ -763,6 +763,29 @@ func (db *DB) Delete(
...
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "DB.RunInTx: cannot start a database transaction")
```
Comment (lippserd) [inline database/db.go:774]: Omit the prefix. Also, `"can't start transaction"` should be totally sufficient because of our error wrapping.

File: database/db.go [hunk 4]
```diff
@@ -763,6 +763,29 @@ func (db *DB) Delete(
...
+	}
+	// We don't expect meaningful errors from rolling back the tx other than the sql.ErrTxDone, so just ignore it.
+	defer func() { _ = tx.Rollback() }()
+
+	if err := fn(ctx, tx); err != nil {
+		return err
+	}
+	if err := tx.Commit(); err != nil {
```
Comment (lippserd) [inline database/db.go:782]: I would add a newline before.

File: database/db.go [hunk 5]
```diff
@@ -763,6 +763,29 @@ func (db *DB) Delete(
...
+func (db *DB) RunInTx(ctx context.Context, fn func(context.Context, *sqlx.Tx) error) error {
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "DB.RunInTx: cannot start a database transaction")
+	}
+	// We don't expect meaningful errors from rolling back the tx other than the sql.ErrTxDone, so just ignore it.
+	defer func() { _ = tx.Rollback() }()
+
+	if err := fn(ctx, tx); err != nil {
+		return err
```
Comment (lippserd) [inline database/db.go:780]: No wrapping here on purpose?

**Review by lippserd — CHANGES_REQUESTED** [review]

File: database/db.go (hunk 5 above)
Comment (yhabteab) [reply database/db.go:780]: What should I wrap it with? Like it states in the method doc block, if the callback returns an error, then it is returned as-is. If in doubt, that provided callback should returned wrapped errors to clearly indicate what's going wrong.

File: database/db.go (hunk 2 above)
Comment (yhabteab) [reply database/db.go:771]: > Please also consider that we already have special handling for serializable transactions.

Which special handling are you referring to? As far as I am aware, our HA tx utilises different method for MySQL driver to simulate a serializable transaction. However, I don't see how this method should make any indication of that.

Comment (lippserd) [reply database/db.go:771]: > However, I don't see how this method should make any indication of that.

Why not? I don't understand your answer to be honest.

File: database/db.go (hunk 5 above)
Comment (lippserd) [reply database/db.go:780]: With a stacktrace.

Comment (lippserd) [reply database/db.go:780]: > What should I wrap it with?

With a stacktrace.

File: database/db.go (hunk 2 above)
Comment (yhabteab) [reply database/db.go:771]: This method only executes the provided callback and never gets to see what queries you are executing. So how should I append `LOCK IN SHARE MODE` to every single query that is executed within that callback?

File: database/db.go (hunk 5 above)
Comment (yhabteab) [reply database/db.go:780]: Just wrapping with `errors.WithStack()` without additional messages? I can do this, but I just wanted to avoid another layer that meaninglessly wraps the error, which is not helpful for debugging and always expect the callback to return wrapped errors *with context specific messages*.

File: database/db.go (hunk 2 above)
Comment (yhabteab) [reply database/db.go:771]: Therefore, I would just allow to pass `sql.TxOptions` as an argument to this method, and the caller has to perform the special cases required for that callback.

File: database/db.go (hunk 5 above)
Comment (yhabteab) [reply database/db.go]: Fixed!

File: database/db.go (hunk 2 above)
Comment (yhabteab) [reply database/db.go]: We'll introduce another version of `ExecTx` in the future that takes care of the tx isolation special handling, but this method won't change as it only invokes the provided callback with a newly initialised tx and has no knowledge of the actual queries executed by `fn`.

**Comment by yhabteab** [comment]
Just fixed the outdated commit message which was referring to the now renamed method `DB#RunInTx`.

File: database/db.go (hunk 5 above)
Comment (lippserd) [reply database/db.go]: We had a conversation about this, and I wanted to add `errors.WithStack()` because otherwise there is no clear indication that the (error from the) callback actually went through this function.

File: database/db.go [hunk 6]
```diff
@@ -769,6 +769,30 @@ func (db *DB) Delete(
 	return db.DeleteStreamed(ctx, entityType, idsCh, onSuccess...)
 }
 
+// ExecTx allows running a function in a database transaction without requiring manual transaction handling.
```
Comment (lippserd) [inline database/db.go:772]: Description would benefit from simple english and simple tenses, e.g. "allows running a function" vs. "executes", "is started" vs "starts".

I would write the following:

ExecTx executes a function within a database transaction.

Starts a new transaction, executes the provided function, and commits the transaction if the function succeeds. If the function returns an error, the transaction is rolled back.

Returns an error if starting the transaction, executing the function, or committing the transaction fails.

**Review by lippserd — CHANGES_REQUESTED** [review]

File: database/db.go [hunk 7]
```diff
@@ -769,6 +769,30 @@ func (db *DB) Delete(
...
+//
+// A new transaction is started on [DB] which is then passed to fn. After fn returns, the transaction is
+// committed unless an error was returned. If fn returns an error, that error is returned or when failing
+// to start or/and commit the transaction.
+func (db *DB) ExecTx(ctx context.Context, fn func(context.Context, *sqlx.Tx) error) error {
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "cannot start transaction")
```
[the remark this answers is not in the document]
Comment (yhabteab) [reply database/db.go:780]: Quite ironic that we have replaced all such abbreviations in Icinga Notifications :-)

[the remark this answers is not in the document]
Comment (lippserd) [reply database/db.go]: I didn't know that :). If we want to use more formal error messages, I have no problem with that. However, we should then change all error messages in the lib.

**Review by lippserd — APPROVED** [review]

**Comment by yhabteab** [comment]
@oxzi do you want to look over the changes again since you previously approved it?

**Review by oxzi — APPROVED** [review]
